### PR TITLE
Plumber

### PR DESCRIPTION
**Pull request by tonyfujs**
Add plumber endpoints

### CODE REVIEW

**Review by Aeilert — COMMENTED** [review]
Hey @tonyfujs, 

Here are some comment on the code that has been added in this branch. 

I look briefly at some on the other existing code as well, but since those files are not part of this PR I haven't made any line number comments. Some general thoughts: 

* Consider using `sprintf` instead of `paste`, e.g. in `get_svy_data()` and `add_dist_stats()` 
* Get rid of `{stringr}`, e.g. in `get_svy_data()`. 
* The `(svy_coverage %in% c("urban", "rural")` statement in `get_svy_data()`  should probably be seen in relation to my comment on the `svy_coverage` parameter. 
* I believe `df[, .SD, .SDcols = vars]` is generally safer than  `df[, ..vars]`, eg. in `collapse_rows()`
* I'm not really sure on the best way to handle the metadata, including collapsing returned rows with a `|` separator and doing a cartesian join. 
* `fg_pip()` and `rg_pip()` seems to reuse some code that probably could be extracted as smaller functions.
* `ag_average_poverty_stats()` does a bit of multiplications that could 1) probably be extracted as separate functions and 2) possibly be included in `{wbpip}`?

File: R/get_aux_table.R [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#' @param data_dir character: Data directory
+#' @param table character: Name of auxiliary table
+#'
+#' @return
+#' @export
+#'
+#' @examples
+get_aux_table <- function(data_dir, table) {
+
+  out <- fst::read_fst(paste0(data_dir, "_aux/", table, ".fst"))
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply R/get_aux_table.R:12]: Good point. I'll use sprintf instead.

File: R/pip.R [hunk 2]
```diff
@@ -1,17 +1,37 @@
...
                 povline   = NULL,
                 popshare  = NULL,
-                year      = "all",
                 fill_gaps = FALSE,
                 aggregate = FALSE,
                 group_by  = NULL,
-                welfare_type = "all",
-                svy_coverage = "all",
+                welfare_type = c("all", "consumption", "income"),
+                svy_coverage = c("all", "rural", "urban"),
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply R/pip.R:27]: That's right, good time to straitghten this up.

File: R/pip.R [hunk 3]
```diff
@@ -1,17 +1,37 @@
...
+#' @param welfare_type character: Welfare type.
+#' @param svy_coverage character: Survey coverage.
+#' @param ppp numeric: Custom Purchase Power Parity value
+#' @param lkup
+#' @param paths
+#'
+#' @return data.frame
+#' @export
+#'
 pip <- function(country   = "all",
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply R/pip.R:19]: We will be using WLD, but still needs to be implemented

File: R/pip.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (tonyfujs) [reply R/pip.R:27]: mmm... I think I disagree with this: Having "aggregated" as an option... What we care about is the "geographical area level" of the statistics to be returned. "aggregated" is mostly related to how the statistics are being computed...

File: R/start_api.R [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+#'
+#' @return
+#' @export
+#'
+start_api <- function(api_version = "v1",
+                      port = 80,
+                      host = "0.0.0.0") {
+  version_path <- paste0("plumber/", api_version, "/plumber.R")
+  api_path <- system.file(version_path, package = "pipapi")
+  api <- plumber::plumb(api_path)
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply R/start_api.R:15]: Not sure either... Maybe we can test and see

File: inst/plumber/v1/plumber.R [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  #             povline      = req$argsQuery$povline,
+  #             popshare     = req$argsQuery$popshare,
+  #             fill_gaps    = req$argsQuery$fill_gaps,
+  #             aggregate    = req$argsQuery$aggregate,
+  #             group_by     = req$argsQuery$group_by,
+  #             welfare_type = req$argsQuery$welfare_type,
+  #             svy_coverage = req$argsQuery$svy_coverage,
+  #             ppp          = req$argsQuery$ppp,
+  #             lkup         = lkups,
+  #             paths        = paths)
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply inst/plumber/v1/plumber.R:114]: Yes!

File: inst/plumber/v1/plumber.R [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+#   pipapi::get_pip_version()
+# }
+
+#* Return CPI table
+#* @get /api/v1/get-cpi
+#* @serializer csv
+function() {
+  out <- pipapi::get_aux_table(data_dir = data_folder_root,
+                               table = "cpi")
+  plumber::as_attachment(out, "cpi.csv")
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply inst/plumber/v1/plumber.R:69]: Yes, there will an additional parameter here top control the return value: json or download

File: inst/plumber/v1/plumber.R [hunk 7]
```diff
@@ -0,0 +1,123 @@
+#* @apiTitle PIP API
+#* @apiVersion 0.0.1
+#* @apiDescription This API powers computations of statistics available at
+#* pip.worldbank.org
+
+library(pipapi)
+
+source("../../../TEMP/TEMP_clean_data.R")
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply inst/plumber/v1/plumber.R:8]: Definitely!

File: inst/plumber/v1/openapi.yaml [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+      responses:
+        default:
+          description: Default response.
+      parameters: []
+  /api/v1/version:
+    get:
+      summary: ' Returns the version to the code and data that powers the API'
+      responses:
+        default:
+          description: Default response.
```
[the remark this answers is not in the document]
Comment (tonyfujs) [reply inst/plumber/v1/openapi.yaml:19]: Good point. Need to dig a bit more on this...

**Pull request merged**